Zipfile bug fix

redruin1 opened the pull request:
Issue being fixed: #78 
Sets the `ZLIB_INCLUDE` and `ZLIB_LIBRARY` paths to the precompiled binary packaged in Overgrowth's libraries, overriding the default one provided by FreeType. I've also set `FT_REQUIRE_ZLIB` to `ON` so that if in the future FreeType is unable to find this precompiled version it will give an error immediately instead of silently and cryptically failing later. Both of these changes are only applied in a Windows build system. Works on Windows 10, and tested on Ubuntu 20.04 LTS with no noticeable changes in the build system from cursory testing. 

Only other changes in this PR are removing some now redundant debug code and changing another engine iterator to a regular for loop (should I create a separate issue to track these?).

gonnavis commented:
Thanks! I succeed with
- This pr.
- CMakeLists.txt changes, which also @redruin1 told me.
![image](https://user-images.githubusercontent.com/10785634/204077198-ec76f930-490c-4ec5-a6cd-c86751c6644e.png)
- cmake gui. ( important! )
- VS2022 Release x64 target.

redruin1 commented:
More on the CMakeLists changes made when discussing with @gonnavis:

There seems to be some issue with Visual Studio when working with the CMake command line that isn't present when working with the GUI. Notably, if you use the GUI to generate the project files from a clean build Overgrowth should compile right out of the box; but if you use the command line it seems to not register the changes to the CMakeLists and fails to compile the fix, resulting in Overgrowth building, but with the zipfile error present again. However, while still in the solution file, if you change any of the CMakeLists (or do anything that forces VS to rerun cmake internally) the solution will prompt you to reload; if you do so, then the fix is compiled and the patch works.

Intuition seems to indicate that this is some kind of caching issue, but this issue persists even if you completely delete the contents of the Build folder in it's entirety; there should be no cache to revert to. It must be that for whatever reason, this patch I've pushed is just not getting run on the first time around when running CMake from the command line; *why* this is the case is anyone's guess.

Not a huge issue by any means; easiest solution right now is to just use the GUI. However, all of the current documentation points to the command line as a valid procedure, so it would be better if this was investigated further as this could be a hallmark of a more insidious issue.

I've also been experimenting with different (perhaps more conventional) ways to specify the zlib path. I changed `ZLIB_INCLUDE` to `ZLIB_INCLUDE_DIR`, as I believe this is more correct, and I got issues with the former when experimenting with clean builds. I also changed the statements to regular `set` commands; It turns out that relative paths are allowed, they just must be prepended with a known directory, like so:
```cmake
IF(WIN32)
    # ...
    SET(ZLIB_INCLUDE_DIR "${PROJECT_SOURCE_DIR}/../Libraries/zlib1.2.7/include")
    SET(ZLIB_LIBRARY "${PROJECT_SOURCE_DIR}/../Libraries/zlib1.2.7/lib/x64/zlib.lib")
ENDIF()
```
This might be better than `get_filename_component`, as that command seems to be somewhat deprecated. Other avenues I explored was a custom [find module](https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html#find-modules), though I honestly don't know enough about cmake to know whether or not that would be a worthwhile investment than just setting the paths directly.

redruin1 commented:
Alright, after *much* research and dozens of from-scratch compilations I've revised my proposed solution to:
```cmake
IF(WIN32)
    # Make sure freetype uses the precompiled 1.2.7 binary
    OPTION(FT_REQUIRE_ZLIB "Overridden on Windows" ON)
    
    SET(ZLIB_INCLUDE_DIR "${PROJECT_SOURCE_DIR}/../Libraries/zlib1.2.7/include" CACHE INTERNAL "Precompiled zlib include path")
    SET(ZLIB_LIBRARY "${PROJECT_SOURCE_DIR}/../Libraries/zlib1.2.7/lib/${WINDOWS_LIB_DIR}/zlib.lib" CACHE INTERNAL "Precompiled zlib library path")
ENDIF()
```
Since `FT_REQUIRE_ZLIB` is specified as an option I mirror that here, since from what I've read this is more correct. Both `set` statements were changed to cache variables (so they're guaranteed to be global and visible on the first run) and labeled them as internal so that the user shouldn't be able to easily change them. Most notably, the library path is now also dependent on the build target being made, either `x86` or `x64`, so users should be able to link with the 32-bit version just fine (which I have confirmed in my testing). Additionally, these changes seem to have resolved the CMake command-line/GUI discrepancy; I built and tested all permutations of command-line/GUI and Release/Debug, and in every case not only was the bug resolved but it worked out of the box first-try.

Due to my research, there is a slight alteration on this CMakeLists that's roughly equivalent that uses a single statement to specify the location of zlib:

```cmake
SET(CMAKE_PREFIX_PATH "${PROJECT_SOURCE_DIR}/../Libraries/zlib1.2.7")
```

This finds the headers just fine, but due to the unconventional structure of the lib folder (two folders, x86 and x64) CMake cannot find the libraries, and has to be specified alongside:

```cmake
SET(CMAKE_PREFIX_PATH "${PROJECT_SOURCE_DIR}/../Libraries/zlib1.2.7")
SET(ZLIB_LIBRARY "${PROJECT_SOURCE_DIR}/../Libraries/zlib1.2.7/lib/${WINDOWS_LIB_DIR}/zlib.lib" CACHE INTERNAL "Precompiled zlib library path")
```

This pair seems to behave identically at least in my testing, but I've opted for the former since it seems more readable and avoids potential clashes if some other part of CMake is using the prefix path. However, if CMake can be informed of this unconventional folder structure, it might be better to switch to this command; but as of writing I have been unable to find reasonable solution aside from specifying the library location manually as above.

shinymerlyn commented:
Re the caching issue -

I pretty much exclusively use cmake from the command line. Sometimes VS sees changes to CMakeLists. In those cases, I run the build, it errors out, I get a reload prompt and reload, then run the build again and it's all good. Occasionally VS doesn't see the change to CMakeLists. But if I nuke the Build subdir and rerun cmake, I never have a "stale" build at that point. Makes sense, since the cache lives in the build folder, and if you nuke it the cache is gone.

redruin1 commented:
@shinymerlyn Actually, the behavior you describe is basically identical to the behavior that Vis and I had: compiles fine first time around but with the bug still there, make some change to CMakeLists so that VS is forced to rerun CMake, a prompt to reload the entire solution comes up, reload all, recompile, and the patch works. You'd think VS would also have a dedicated "Reload solution" button, but from what I've found that doesn't seem to be the case. Perfectly surmountable, provided you know what to do; but figuring out what to do could be near impossible to the untrained, and there's no documentation regarding it, so I wanted to see if it was something I could resolve with the build itself.

I think the issue was caused by the set command being local to the CMakeLists it's specified in; if you delete the `CACHE INTERNAL ...` and just have it be a regular `set` statement, nuke the build folder and recompile, you'll actually get configuration errors (or, at least I did) where CMake can't resolve the paths set. (Actually, more confusingly, it finds zlib, and then immediately *can't* find zlib, resulting in an un-googleable error string. Probably because I'm manually setting the ZLIB_XXX variables which are normally found by a CMake config/find module.) I *suspect* because when CMake travels into the freetype subdirectory it travels out of the variables scope and can't be recognized; but when you set it to a cache variable it essentially becomes global and can be found across the solution. Of course, this doesn't explain why it fails to work the first time around, and then succeeds the second time around... it's all very cryptic.
